### ISSUE TITLE
speeding up local queries? 

### ISSUE THREAD

**Issue by jtmiller28** [cboettig/gbifdb#12]
Hello, 
I was wondering if there is anyway to speed up local queries for gbifdb? Currently using collect() requires a large amount of time to process data, too much for our lab when working at scale. 

e.g. 
`### Download Script for GBIF db 
library(gbifdb) # load dep
library(dplyr)
library(tidyverse)
Sys.setenv(GBIF_HOME = "/blue/soltis/millerjared/BoCP/data/gbif-db/")
#gbif_download(dir = gbif_dir("/blue/soltis/millerjared/BoCP/data/gbif-db/"))

gbif <- gbif_local() 
gbif

time_taken <- system.time({
check_df <- gbif %>% 
  filter(species == "Ursus americanus") %>% 
  count(scientificname)

check_df <- check_df %>% collect()

})

elapsed_minutes <- time_taken["elapsed"] / 60`

elapsed_minutes = 2.49 which when ran at scale of >30,000 species is way too long: ~1,365 hours

My advisor and I discussed parsing down the parquet files to remove birds from the data, but I am pretty sure that the issue is more on the collect() end where the data is pulled into R's working memory. 

Is there any way to speed this process up? I could run it in parallel, but thats a lot of CPUs so I wanted to check first. 

Session Info: 
R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.4 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

time zone: America/New_York
tzcode source: system (glibc)

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] lubridate_1.9.3 forcats_1.0.0   stringr_1.5.1   purrr_1.0.2    
 [5] readr_2.1.5     tidyr_1.3.1     tibble_3.2.1    ggplot2_3.5.1  
 [9] tidyverse_2.0.0 dplyr_1.1.4     gbifdb_1.0.0   

Hardware Resource Request Info: 
Rstudio IDE (not ran as a script through slurm scheduler) 
6 CPUs 
32 GB RAM 

Thanks! 
JT

**Comment by cboettig** [comment]
@jtmiller28 thanks for asking!  yes, this can indeed be greatly sped up.  For instance, to how many occurrences of each of the over 100,000 species of chordates we might do:

```
bench::bench_time({
  x <- gbif |> filter(phylum=="Chordata") |> count(species) |> collect()
})
```

This takes 5.24 seconds on my machine.  The important thing to realize here is to use dplyr to its fullest advantage, which allows dbplyr to translate your calls into SQL before they are run in collect.  i.e. note that we don't *loop* over each species with filter!  if you want to work species by species, or any other grouping operation, you should learn the pattern of `group_by()` + `summarise()` in dplyr.  `count()` is really just a special case of this, equivalent to 

```
  x <- gbif |> filter(phylum=="Chordata") |> group_by(species) |> summarise(n = n()) |> collect()
```

where `n()` is the count function

**Comment by jtmiller28** [comment]
Oh thanks for that! I default to looping...which is not ideal. 
Is the speed at which dplyr runs often machine dependent? Running the same code: 
`bench::bench_time({
  x <- gbif |> filter(phylum=="Chordata") |> count(species) |> collect()`
takes 1.77 minutes on my IDE interface hosted on the cluster, which is very different than your 5.24 seconds.

**Comment by cboettig** [comment]
yup, the speed at which almost any code runs will vary depending on hardware.  Any hardware component (CPU, RAM, network, disk) can differ in speed, but which one of those is the bottleneck varies by application.   I'm almost certain in this case the difference is disk speed (recall disks have the greatest capacity but also the slowest input/output operations).  I would speak to your cluster admin about this, some clusters have paths that are backed by faster disks.  Also, the 'local' disk space on many clusters is often not actually 'local' but virtual, which depending on configuration can really slow things down.  

Anyway really good questions; I need to figure out how to better communicate these nuances in the docs so it's helpful to have your experiences, which are probably typical of other users as well.  regards!

**Comment by jtmiller28** [comment]
Gotcha, yeah so I talked to my admin and it turns out it gets rather complicated depending on when and where the data was last written for disk allocation. Luckily he showed me a trick of copying it all to the jobs temporary flash storage, then running the local db from there. I was able to get to 28 seconds for: 
`bench::bench_time({ x <- gbif |> filter(phylum=="Chordata") |> count(species) |> collect()`
which is much faster.
